File: slash_commands/streak.py
#!/usr/bin/env python3
import boto3
import discord
from datetime import datetime, timedelta
from random_unicode_emoji import random_emoji
from modules.random_discord_emoji import random_discord_emoji

import megacoin


async def streak(ctx, stats):

    embed = discord.Embed(color=0xfee9b6,
                          title="⏳  Loading...")

    interaction = await ctx.respond(embed=embed)

    TABLE = "discord-streak"
    ddb = boto3.resource('dynamodb')
    table = ddb.Table(TABLE)

    dataStats = table.get_item(
        Key={
            'id': 'allTimeStreak'
        }
    )

    dataCurrent = table.get_item(
        Key={
            'id': 'currentStreak'
        }
    )

    data = table.get_item(
        Key={
            'id': str(ctx.user.id)
        }
    )

    dataLength = int(data['ResponseMetadata']
                     ['HTTPHeaders']['content-length'])

    # Format: 2023-02-22 21:11:05.800895
    if dataLength > 5:
        storedLastMid = datetime.fromisoformat(data['Item']['lastMid'])
        storedNextMid = datetime.fromisoformat(data['Item']['nextMid'])
        storedSkipMid = datetime.fromisoformat(data['Item']['skipMid'])

    prefix = ""
    streak = 0
    personalRecord = 0
    hit = False

    currTime = datetime.now()
    lastMidnight = datetime.combine(currTime.date(), datetime.min.time())
    nextMidnight = lastMidnight + timedelta(days=1)
    skipMidnight = lastMidnight + timedelta(days=2)

    if dataLength < 5:
        prefix = "You just started a new streak! "
        hit = True

        table.put_item(
            Item={
                'id': str(ctx.user.id),
                'username': str(ctx.user.name),
                'streak': str(int(streak) + 1),
                'updated': str(currTime),
                'lastMid': str(lastMidnight),
                'nextMid': str(nextMidnight),
                'skipMid': str(skipMidnight),
                'personalRecord': str(int(personalRecord) + 1)
            }
        )

        streak = streak + 1

    elif currTime > storedSkipMid:
        prefix = "You missed your streak! "
        hit = True

        table.put_item(
            Item={
                'id': str(ctx.user.id),
                'username': str(ctx.user.name),
                'streak': str(int(streak) + 1),
                'updated': str(currTime),
                'lastMid': str(lastMidnight),
                'nextMid': str(nextMidnight),
                'skipMid': str(skipMidnight),
                'personalRecord': str(data['Item']['personalRecord'])
            }
        )

        streak = streak + 1

        if str(ctx.user.id) == str(dataCurrent['Item']['userId']):
            table.put_item(
                Item={
                    'id': 'currentStreak',
                    'stat': '0',
                    'username': str(ctx.bot.user.name),
                    'userId': str(ctx.bot.user.id)
                }
            )

            prefix = prefix + "**Current Highest Streak Reset!** "

    elif currTime > storedLastMid and currTime < storedNextMid:
        prefix = ""

        streak = int(data['Item']['streak'])

        table.put_item(
            Item={
                'id': str(ctx.user.id),
                'username': str(ctx.user.name),
                'streak': str(streak),
                'updated': str(currTime),
                'lastMid': str(data['Item']['lastMid']),
                'nextMid': str(data['Item']['nextMid']),
                'skipMid': str(data['Item']['skipMid']),
                'personalRecord': str(data['Item']['personalRecord'])
            }
        )

    elif currTime > storedLastMid and currTime > storedNextMid:
        prefix = "You hit your streak! "
        hit = True
        streak = int(data['Item']['streak']) + 1
        personalRecord = int(data['Item']['personalRecord'])

        if streak > personalRecord:
            personalRecord = streak

        table.put_item(
            Item={
                'id': str(ctx.user.id),
                'username': str(ctx.user.name),
                'streak': str(streak),
                'updated': str(currTime),
                'lastMid': str(lastMidnight),
                'nextMid': str(nextMidnight),
                'skipMid': str(skipMidnight),
                'personalRecord': str(personalRecord)
            }
        )

        if streak > int(dataStats['Item']['stat']):
            table.put_item(
                Item={
                    'id': 'allTimeStreak',
                    'stat': str(streak),
                    'username': str(ctx.user.name),
                    'userId': str(ctx.user.id)
                }
            )

        if streak > int(dataCurrent['Item']['stat']):
            table.put_item(
                Item={
                    'id': 'currentStreak',
                    'stat': str(streak),
                    'username': str(ctx.user.name),
                    'userId': str(ctx.user.id)
                }
            )

    if streak < 2:
        emote = "💩"
    elif streak < 3:
        emote = "✌️"
    elif streak < 4:
        emote = "👌"
    elif streak < 5:
        emote = "🍀"
    elif streak < 10:
        emote = "🔥"
    elif streak < 25:
        emote = "🧨"
    elif streak < 50:
        emote = "🏆"
    elif streak < 69:
        emote = "💀"
    elif streak < 70:
        emote = "-  *nice*."
    elif streak < 75:
        emote = "💀"
    elif streak < 100:
        emote = "💎"
    elif streak < 101:
        emote = "💯 - *Welcome to Party Mode*"
    else:
        guild = discord.utils.get(ctx.bot.guilds, name="MEGACORD")
        emojis = await guild.fetch_emojis()
        for _ in range(0, 5):
            emojis = emojis + emojis
        emote = str(await random_discord_emoji(guild, ctx.bot, "MEGACORD"))

    statMessage = ""

    if stats:
        statMessage = "\n\n📊\n**All-Time Highest Streak:** " + \
            str(dataStats['Item']['stat']) + \
            " *by* <@" + \
            str(dataStats['Item']['userId']) + \
            ">" + \
            "\n**Current Highest Streak:** " + \
            str(dataCurrent['Item']['stat']) + \
            " *by* <@" + \
            str(dataCurrent['Item']['userId']) + \
            ">" + \
            "\n**Personal Highest Streak:** " + \
            str(data['Item']['personalRecord']) + \
            " *by* <@" + \
            str(ctx.user.id) + ">"

    content = prefix + "Your streak is: " + \
        str(streak) + "  " + emote

    color = 0x5965f3

    if "missed" in prefix:
        color = 0xdd2f45

    embed = discord.Embed(color=color, title=content, description=statMessage)
    bonusText = ""
    bonus = 0

    if hit:
        coins = streak
        if coins > 100:
            coins = 100
        if (streak % 100) == 0:
            bonus = streak
            bonusText = f"  + {bonus} bonus <:MEGACOIN:1090620048621707324>"
        embed = discord.Embed(color=color, title=content,
                              description=f"+ {coins} <:MEGACOIN:1090620048621707324>{bonusText}\n{statMessage}")

        coins += bonus
        await megacoin.add(ctx.user, coins)

    await interaction.edit_original_response(embed=embed)

File: services/streak.ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
} from "@aws-sdk/lib-dynamodb";

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

const STREAK_TABLE = process.env.STREAK_TABLE || "discord-streak";

export interface StreakData {
  id: string;
  username: string;
  streak: number;
  updated: string;
  lastMid: string;
  nextMid: string;
  skipMid: string;
  personalRecord: number;
}

export interface StreakStats {
  allTimeStreak: { stat: number; username: string; userId: string };
  currentStreak: { stat: number; username: string; userId: string };
}

/**
 * Get a user's streak data
 */
export async function getStreakData(
  userId: string,
): Promise<StreakData | null> {
  const response = await docClient.send(
    new GetCommand({
      TableName: STREAK_TABLE,
      Key: { id: userId },
    }),
  );

  if (!response.Item) {
    return null;
  }

  return {
    id: response.Item.id as string,
    username: response.Item.username as string,
    streak: parseInt(response.Item.streak as string, 10) || 0,
    updated: response.Item.updated as string,
    lastMid: response.Item.lastMid as string,
    nextMid: response.Item.nextMid as string,
    skipMid: response.Item.skipMid as string,
    personalRecord: parseInt(response.Item.personalRecord as string, 10) || 0,
  };
}

/**
 * Get streak stats (all-time and current)
 */
export async function getStreakStats(): Promise<StreakStats> {
  const [allTimeResponse, currentResponse] = await Promise.all([
    docClient.send(
      new GetCommand({
        TableName: STREAK_TABLE,
        Key: { id: "allTimeStreak" },
      }),
    ),
    docClient.send(
      new GetCommand({
        TableName: STREAK_TABLE,
        Key: { id: "currentStreak" },
      }),
    ),
  ]);

  return {
    allTimeStreak: {
      stat: parseInt(allTimeResponse.Item?.stat as string, 10) || 0,
      username: (allTimeResponse.Item?.username as string) || "Unknown",
      userId: (allTimeResponse.Item?.userId as string) || "",
    },
    currentStreak: {
      stat: parseInt(currentResponse.Item?.stat as string, 10) || 0,
      username: (currentResponse.Item?.username as string) || "Unknown",
      userId: (currentResponse.Item?.userId as string) || "",
    },
  };
}

/**
 * Update a user's streak
 */
export async function updateStreak(
  userId: string,
  username: string,
  streak: number,
  personalRecord: number,
  lastMid: Date,
  nextMid: Date,
  skipMid: Date,
): Promise<void> {
  await docClient.send(
    new PutCommand({
      TableName: STREAK_TABLE,
      Item: {
        id: userId,
        username: username,
        streak: streak.toString(),
        updated: new Date().toISOString(),
        lastMid: lastMid.toISOString(),
        nextMid: nextMid.toISOString(),
        skipMid: skipMid.toISOString(),
        personalRecord: personalRecord.toString(),
      },
    }),
  );
}

/**
 * Update streak stats
 */
export async function updateStreakStats(
  type: "allTimeStreak" | "currentStreak",
  stat: number,
  username: string,
  userId: string,
): Promise<void> {
  await docClient.send(
    new PutCommand({
      TableName: STREAK_TABLE,
      Item: {
        id: type,
        stat: stat.toString(),
        username: username,
        userId: userId,
      },
    }),
  );
}

/**
 * Process a streak hit and return the result
 */
export async function processStreak(
  userId: string,
  username: string,
  botUserId: string,
  botUsername: string,
): Promise<{
  streak: number;
  personalRecord: number;
  prefix: string;
  hit: boolean;
  isNewRecord: boolean;
}> {
  const now = new Date();
  const lastMidnight = new Date(now);
  lastMidnight.setHours(0, 0, 0, 0);

  const nextMidnight = new Date(lastMidnight);
  nextMidnight.setDate(nextMidnight.getDate() + 1);

  const skipMidnight = new Date(lastMidnight);
  skipMidnight.setDate(skipMidnight.getDate() + 2);

  const existingData = await getStreakData(userId);
  const stats = await getStreakStats();

  let streak = 0;
  let personalRecord = 0;
  let prefix = "";
  let hit = false;
  let isNewRecord = false;

  if (!existingData) {
    // New user starting a streak
    prefix = "You just started a new streak! ";
    hit = true;
    streak = 1;
    personalRecord = 1;
    isNewRecord = true;

    await updateStreak(
      userId,
      username,
      streak,
      personalRecord,
      lastMidnight,
      nextMidnight,
      skipMidnight,
    );
  } else {
    const storedSkipMid = new Date(existingData.skipMid);
    const storedNextMid = new Date(existingData.nextMid);
    const storedLastMid = new Date(existingData.lastMid);

    if (now > storedSkipMid) {
      // Missed streak
      prefix = "You missed your streak! ";
      hit = true;
      streak = 1;
      personalRecord = existingData.personalRecord;

      await updateStreak(
        userId,
        username,
        streak,
        personalRecord,
        lastMidnight,
        nextMidnight,
        skipMidnight,
      );

      // Reset current streak if they were the leader
      if (stats.currentStreak.userId === userId) {
        await updateStreakStats("currentStreak", 0, botUsername, botUserId);
        prefix += "**Current Highest Streak Reset!** ";
      }
    } else if (now > storedLastMid && now < storedNextMid) {
      // Already hit today
      prefix = "";
      streak = existingData.streak;
      personalRecord = existingData.personalRecord;
    } else if (now >= storedNextMid && now < storedSkipMid) {
      // Valid streak continuation (matches Python: currTime > storedLastMid and currTime > storedNextMid)
      prefix = "You hit your streak! ";
      hit = true;
      streak = existingData.streak + 1;
      personalRecord = Math.max(existingData.personalRecord, streak);
      isNewRecord = streak > existingData.personalRecord;

      await updateStreak(
        userId,
        username,
        streak,
        personalRecord,
        lastMidnight,
        nextMidnight,
        skipMidnight,
      );

      // Update current streak if this is the new high
      if (streak > stats.currentStreak.stat) {
        await updateStreakStats("currentStreak", streak, username, userId);
      }

      // Update all-time streak if this is the new record (Python doesn't add special text)
      if (streak > stats.allTimeStreak.stat) {
        await updateStreakStats("allTimeStreak", streak, username, userId);
      }
    } else {
      // Edge case: before lastMid
      prefix = "";
      streak = existingData.streak;
      personalRecord = existingData.personalRecord;
    }
  }

  return { streak, personalRecord, prefix, hit, isNewRecord };
}
